File: ios/ScreenTimeManager.m
//
//  ScreenTimeManager.m
//  appblocker
//
//  Created by Arun Saini on 24/04/26.
//

#import <Foundation/Foundation.h>
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(ScreenTimeManager, NSObject)

RCT_EXTERN_METHOD(requestPermission)
RCT_EXTERN_METHOD(openAppPicker)
RCT_EXTERN_METHOD(startMonitoring:(nonnull NSNumber *)seconds)
RCT_EXTERN_METHOD(unlockApps)
RCT_EXTERN_METHOD(openUnlockScreen)
RCT_EXTERN_METHOD(setDifficulty:(NSString *)level)
RCT_EXTERN_METHOD(stopBlocking)

@end

File: ios/appblocker/ScreenTimeManager.swift
import Foundation
import FamilyControls
import DeviceActivity
import ManagedSettings
import SwiftUI

@objc(ScreenTimeManager)
class ScreenTimeManager: NSObject {

    let store = ManagedSettingsStore()

    static func moduleName() -> String! {
        return "ScreenTimeManager"
    }

    static func requiresMainQueueSetup() -> Bool {
        return true
    }

    // 🔐 Permission
    @objc
    func requestPermission() {
        Task {
            try? await AuthorizationCenter.shared.requestAuthorization(for: .individual)
        }
    }

    // 📱 Picker
    @objc
    func openAppPicker() {
        DispatchQueue.main.async {
            let vc = UIHostingController(rootView: AppPickerView())
            UIApplication.shared.windows.first?.rootViewController?.present(vc, animated: true)
        }
    }

    // 🚀 START TIMER SESSION
    @objc
    func startMonitoring(_ seconds: NSNumber) {

        let duration = seconds.intValue

        let now = Date()
        let end = Calendar.current.date(byAdding: .second, value: duration, to: now)!

        let start = Calendar.current.dateComponents([.hour, .minute, .second], from: now)
        let finish = Calendar.current.dateComponents([.hour, .minute, .second], from: end)

        let schedule = DeviceActivitySchedule(
            intervalStart: start,
            intervalEnd: finish,
            repeats: false
        )

        try? DeviceActivityCenter().startMonitoring(
            DeviceActivityName("session"),
            during: schedule
        )
    }

    // 🔓 Unlock
    @objc
    func unlockApps() {
        store.shield.applications = nil
    }
    @objc
func openUnlockScreen() {
    DispatchQueue.main.async {
        let vc = UIHostingController(rootView: UnlockView())
        UIApplication.shared.windows.first?.rootViewController?.present(vc, animated: true)
    }
}
}